### DISCUSSION TITLE
Is hardware accelerated tonemapping available for Intel 9th-gen through jellyfin-docker?

### DISCUSSION THREAD

**Discussion by henkisdabro** [jellyfin/jellyfin#6373] [Q&A] [answered]
I have read through the [documentation for quicksync hardware acceleration](https://jellyfin.org/docs/general/administration/hardware-acceleration.html#intel-quicksync) and have it working properly via the jellyfin-docker linuxserver.io image (after curling home an updated version of jellyfin-ffmpeg). I'm running an i5-9400T CPU which in theory should support hardware tonemapping (?), but I see on[ this section of the documentation](https://jellyfin.org/docs/general/administration/hardware-acceleration.html#configuring-opencl-acceleratedvpp-tone-mapping) that it might not work over docker - can someone confirm if this is the case and whether my CPU actually can support hw tonemapping if I were to install jellyfin on bare metal Debian instead?

**Comment by nyanmisaka** [comment, marked as answer]
Sure. But you need the `intel-media-va-driver-non-free` 20.1 or newer installed on your Debian host.
Older versions of intel driver doesn't support this function.

**Reply by henkisdabro** [reply]
Thanks for your reply! Do you mean that it works also on Jellyfin via Docker or does it only work on bare metal installs?

**Reply by nyanmisaka** [reply]
It works on bare metal installs but you have to install Intel drivers by youself.

**Reply by henkisdabro** [reply]
Thanks! Just to confirm - are we sure it does NOT work on docker Jellyfin?

**Reply by crobibero** [reply]
It will also work in docker, but you have to manually install the drivers

**Reply by henkisdabro** [reply]
Great, thanks for this confirmation! I have installed both ffmpeg and the intel driver into my container and also set the container to privileged mode as advised in the docementation, but I'm getting the following error messages when trying to play movies (I have ticked all boxes in the hardware acceleration section of the jellyfin UI (including the tone mapping setting).

Here are the error messages from the docker logs for jellyfin-docker:


```
[20:43:45] [INF] [40] Jellyfin.Api.Helpers.TranscodingJobHelper: /usr/lib/jellyfin-ffmpeg/ffmpeg -hwaccel vaapi -init_hw_device vaapi=va:/dev/dri/renderD128 -init_hw_device qsv@va -hwaccel_output_format vaapi -i file:"/data/movies/TestingMovie2160pHDR/TestingMovie2160pHDR.mkv" -map_metadata -1 -map_chapters -1 -threads 0 -map 0:0 -map 0:1 -map -0:s -codec:v:0 h264_qsv -preset fast -look_ahead 0 -b:v 30955963 -maxrate 30955963 -bufsize 61911926 -profile:v:0 high -level 41 -g:v:0 72 -keyint_min:v:0 72 -sc_threshold:v:0 0 -vf "tonemap_vaapi=format=nv12:transfer=bt709:matrix=bt709:primaries=bt709,hwmap=derive_device=qsv,scale_qsv" -start_at_zero -vsync -1 -codec:a:0 aac -ac 6 -ab 640000 -copyts -avoid_negative_ts disabled -max_muxing_queue_size 2048 -f hls -max_delay 5000000 -hls_time 3 -hls_segment_type mpegts -start_number 0 -hls_segment_filename "/config/data/transcodes/5a7fb4451d11b513c75953403d11f977%d.ts" -hls_playlist_type vod -hls_list_size 0 -y "/config/data/transcodes/5a7fb4451d11b513c75953403d11f977.m3u8"

[20:43:45] [ERR] [42] Jellyfin.Api.Helpers.TranscodingJobHelper: FFmpeg exited with code 1

[20:43:45] [WRN] [45] Jellyfin.Api.Controllers.DynamicHlsController: cannot serve /config/data/transcodes/5a7fb4451d11b513c75953403d11f9770.ts as transcoding quit before we got there

[20:43:45] [ERR] [45] Jellyfin.Server.Middleware.ExceptionMiddleware: Error processing request: Could not find file '/config/data/transcodes/5a7fb4451d11b513c75953403d11f9770.ts'. URL GET /videos/c23c0df1-e3c5-8521-53c1-406d4da1f8ce/hls1/main/0.ts.

[20:43:50] [INF] [42] Jellyfin.Api.Controllers.DynamicHlsController: Current HLS implementation doesn't support non-keyframe breaks but one is requested, ignoring that request
```

Here are the transcode logs from within jellyfin:
```
Stream mapping:
  Stream #0:0 -> #0:0 (hevc (native) -> h264 (h264_qsv))
  Stream #0:1 -> #0:1 (eac3 (native) -> aac (native))
Press [q] to stop, [?] for help
[Parsed_tonemap_vaapi_0 @ 0x5654bc546800] No mastering display data from input
Error while filtering: Invalid argument
Failed to inject frame into filter network: Invalid argument
Error while processing the decoded data for stream #0:0
[aac @ 0x5654bc3d8ac0] Qavg: 65536.000
[aac @ 0x5654bc3d8ac0] 2 frames left in the queue on closing
Conversion failed!
```



Here are the commands I've issued to install the above:
```
apt update
curl -LO https://repo.jellyfin.org/releases/server/ubuntu/versions/jellyfin-ffmpeg/4.3.2-1/jellyfin-ffmpeg_4.3.2-1-focal_amd64.deb
dpkg -i jellyfin-ffmpeg_4.3.2-1-focal_amd64.deb
apt install vainfo intel-media-va-driver-non-free -y
```

Am I missing something, I would really appreciate your help!

**Reply by nyanmisaka** [reply]
```[Parsed_tonemap_vaapi_0 @ 0x5654bc546800] No mastering display data from input```

Can you confirm if the media has valid HDR/PQ metadata? Such as smpte2084/smpte2086.

And try a HDR demo from here.
https://4kmedia.org/sony-food-fizzle-uhd-hdr-4k-demo/

**Reply by henkisdabro** [reply]
I downloaded the food-sizzle file you recommended and played it successfully, with tonemapping working, thank you for the guidance! I opened it with the windows app MediaInfo to look at the video metadata and could see the SMPTE2086 HDR tag there, this is great learning! 

The first video file I mentioned in my error logs above still does not work (when transcoding), but as you rightly said it does not have the proper SMPTE2086 tag on it - it seems ffmpeg tries to tonemap it anyways and fail... what can I do to troubleshoot the next steps on this file? Secondly, I have other movie files in my library which _have_ the SMPTE2086 tag and when I try to transcode them, I get choppy playback, or rather video that plays for like 5 seconds fine, then freezes for a second then continues... Any ideas what might cause this type of problem (I'm using the system/boot NVME PCIe SSD for jellyfin transcode cache btw).?

**Reply by nyanmisaka** [reply]
Which HWA method are you using? VAAPI or QSV.

**Reply by nyanmisaka** [reply]
Can you run this command with the first video file and attach the output to here?

```
ffprobe -i "/videos/yourFirstVideo.mkv" -hide_banner -loglevel warning -select_streams v -print_format json -show_frames -show_streams -read_intervals "%+#1" -show_entries "frame=color_space,color_primaries,color_transfer,side_data_list,pix_fmt"
```

**Reply by henkisdabro** [reply]
I'm using the QSV method in Jellyfin -  should I switch to VAAPI?

Here is the output for the first file as requested - I really appreciate your time helping on this!
```
{
    "frames": [
        {
            "pix_fmt": "yuv420p10le",
            "color_space": "bt2020nc",
            "color_primaries": "bt2020",
            "color_transfer": "smpte2084",
            "side_data_list": [
                {
                    "side_data_type": "Content light level metadata",
                    "max_content": 204,
                    "max_average": 55
                }
            ]
        }
    ],
    "streams": [
        {
            "index": 0,
            "codec_name": "hevc",
            "codec_long_name": "H.265 / HEVC (High Efficiency Video Coding)",
            "profile": "Main 10",
            "codec_type": "video",
            "codec_time_base": "1001/24000",
            "codec_tag_string": "[0][0][0][0]",
            "codec_tag": "0x0000",
            "width": 3840,
            "height": 2160,
            "coded_width": 3840,
            "coded_height": 2160,
            "closed_captions": 0,
            "has_b_frames": 2,
            "sample_aspect_ratio": "1:1",
            "display_aspect_ratio": "16:9",
            "pix_fmt": "yuv420p10le",
            "level": 150,
            "color_range": "tv",
            "color_space": "bt2020nc",
            "color_transfer": "smpte2084",
            "color_primaries": "bt2020",
            "refs": 1,
            "r_frame_rate": "24000/1001",
            "avg_frame_rate": "24000/1001",
            "time_base": "1/1000",
            "start_pts": 0,
            "start_time": "0.000000",
            "nb_read_frames": "1",
            "disposition": {
                "default": 1,
                "dub": 0,
                "original": 0,
                "comment": 0,
                "lyrics": 0,
                "karaoke": 0,
                "forced": 0,
                "hearing_impaired": 0,
                "visual_impaired": 0,
                "clean_effects": 0,
                "attached_pic": 0,
                "timed_thumbnails": 0
            },
            "tags": {
                "BPS-eng": "17803251",
                "DURATION-eng": "02:05:25.560000000",
                "NUMBER_OF_FRAMES-eng": "180433",
                "NUMBER_OF_BYTES-eng": "16747429692",
                "_STATISTICS_WRITING_APP-eng": "mkvmerge v50.0.0 ('Awakenings') 64-bit",
                "_STATISTICS_WRITING_DATE_UTC-eng": "2021-02-12 17:08:11",
                "_STATISTICS_TAGS-eng": "BPS DURATION NUMBER_OF_FRAMES NUMBER_OF_BYTES"
            }
        }
    ]
}
```

**Reply by henkisdabro** [reply]
Can i provide anything else to help, just let me know!

**Reply by nyanmisaka** [reply]
This media contains no mastering display data required by the` tonemap_vaapi` filter.
https://github.com/jellyfin/jellyfin-ffmpeg/blob/cec5a97aac350ebd458d5b66cfb2a5c56ea2f110/libavfilter/vf_tonemap_vaapi.c#L123

But only have content light data.
```
            "side_data_list": [
                {
                    "side_data_type": "Content light level metadata",
                    "max_content": 204,
                    "max_average": 55
                }
```

You can use this custom build of ffmpeg to insert a custom mastering data to this media without re-encoding.
https://trac.ffmpeg.org/ticket/7756

**Reply by henkisdabro** [reply]
wow, thanks for the research, I will try to dive in to fully understand your answer :-). I expect more video files will have the problem of missing mastering data, so the solution of using a custom build would solve that each time, or would I need to do any manual file editing before it will work? Secondly I'm not too well-versed with how to install the custom build or how to replace the ffmpeg-jellyfin package that is already installed in my jellyfin container, would you be able to point me in the right direction? Thanks again!

**Comment by nyanmisaka** [comment]
`[Parsed_tonemap_vaapi_0 @ 0x5654bc546800] No mastering display data from input`

This issue can be solved by https://github.com/jellyfin/jellyfin-ffmpeg/pull/73

**Reply by henkisdabro** [reply]
Thanks, that's great! I've tried to figure our how to build a `curl` command to download the `focal` `amd64` build artifact available in that PR, but can't get it to work, is this the only way to get access to the actual build?
